From fc5e376b3b0253f17d44bf5f2347e85d82a78457 Mon Sep 17 00:00:00 2001
From: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
Date: Mon, 5 May 2025 18:03:10 +0200
Subject: [PATCH] lxc/conf: support flag kind of mount options in
 lxc.mount.entry options

Currently, if user wants to use a flag-like mount option in lxc.mount.entry,
for example "userxattr" with overlayfs then it will be silently ignored.

Let's fix that by making parse_vfs_attr() to process all mount options.

Signed-off-by: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
---
 src/lxc/conf.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/src/lxc/conf.c b/src/lxc/conf.c
index f8ab10ce16..04d70a5163 100644
--- a/src/lxc/conf.c
+++ b/src/lxc/conf.c
@@ -2034,6 +2034,11 @@ int parse_mntopts_legacy(const char *mntopts, unsigned long *mntflags, char **mn
 	return 0;
 }
 
+/* Parses generic VFS option (opt) and puts results into (opts).
+ *
+ * @return  1 if opt is not a generic VFS option
+ * @return  0 if opt was processed
+ */
 static int parse_vfs_attr(struct lxc_mount_options *opts, char *opt, size_t size)
 {
 	/*
@@ -2090,7 +2095,8 @@ static int parse_vfs_attr(struct lxc_mount_options *opts, char *opt, size_t size
 		return 0;
 	}
 
-	return 0;
+	/* opt was skipped */
+	return 1;
 }
 
 int parse_mount_attrs(struct lxc_mount_options *opts, const char *mntopts)
@@ -2116,22 +2122,22 @@ int parse_mount_attrs(struct lxc_mount_options *opts, const char *mntopts)
 		return ret_errno(ENOMEM);
 
 	lxc_iterate_parts(mntopt_cur, mntopts_dup, ",") {
-		/* This is a filesystem specific option. */
-		if (strchr(mntopt_cur, '=')) {
-			if (!end) {
-				end = stpcpy(mntopts_new, mntopt_cur);
-			} else {
-				end = stpcpy(end, ",");
-				end = stpcpy(end, mntopt_cur);
-			}
-
-			continue;
-		}
-
 		/* This is a generic vfs option. */
 		ret = parse_vfs_attr(opts, mntopt_cur, size);
 		if (ret < 0)
 			return syserror("Failed to parse mount attributes: \"%s\"", mntopt_cur);
+
+		/* (mntopt_cur) was processed in parse_vfs_attr() */
+		if (ret == 0)
+			continue;
+
+		/* This is a filesystem specific option. */
+		if (!end) {
+			end = stpcpy(mntopts_new, mntopt_cur);
+		} else {
+			end = stpcpy(end, ",");
+			end = stpcpy(end, mntopt_cur);
+		}
 	}
 
 	if (*mntopts_new)
